# Часть 1: Задания, решаемые без использования Python

## Задание 1
Исследуйте систему векторов на линейную зависимость или независимость без использования Python (вручную):

а1 = (1, 8, -1)
а2 =  (-2, 3, 3)
a3 = (4, -11, 9)  

Подробно распишите решение, а в итоге сделайте вывод про то, являются ли векторы зависимыми или независимыми

***Определение линейной зависимости***:

Векторы $a_1, a_2, a_3$ линейно зависимы, если существуют такие коэффициенты $c_1, c_2, c_3$, не все из которых равны нулю, что выполняется:

$
c_1 a_1 + c_2 a_2 + c_3 a_3 = 0.
$

Если таких коэффициентов нет, то векторы линейно независимы.

---
***1. Составим уравнения:***

Подставим векторы $a_1, a_2, a_3$:

$
c_1 (1, 8, -1) + c_2 (-2, 3, 3) + c_3 (4, -11, 9) = (0, 0, 0).
$

Получим систему линейных уравнений:

$
\begin{aligned}
& \quad c_1 - 2c_2 + 4c_3 = 0, \\
& \quad 8c_1 + 3c_2 - 11c_3 = 0, \\
& \quad -c_1 + 3c_2 + 9c_3 = 0.
\end{aligned}
$


***2. Запишем систему в матричном виде:***

$$
\begin{bmatrix}
1 & -2 & 4 \\
8 & 3 & -11 \\
-1 & 3 & 9
\end{bmatrix}
\cdot 
\begin{bmatrix}
c_1 \\
c_2 \\
c_3
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
0
\end{bmatrix}
$$

Чтобы определить, существуют ли нетривиальные решения ($c_1, c_2, c_3$ не равные нулю), нужно найти ранг матрицы

***3. Находим ранг матрицы***
Найдем определитель:

$
\text{det}
\begin{bmatrix}
1 & -2 & 4 \\
8 & 3 & -11 \\
-1 & 3 & 9
\end{bmatrix}
$


$$
\text{det} =
1 \cdot
\begin{vmatrix}
3 & -11 \\
3 & 9
\end{vmatrix}
-
(-2) \cdot
\begin{vmatrix}
8 & -11 \\
-1 & 9
\end{vmatrix}
+
4 \cdot
\begin{vmatrix}
8 & 3 \\
-1 & 3
\end{vmatrix}
$$

$
\text{det} = 1 \cdot 60 - (-2) \cdot 61 + 4 \cdot 27 = 60 + 122 + 108 = 290
$


Так как определитель матрицы $ \neq 0 $, ранг матрицы равен 3 (максимальный). Следовательно, система уравнений имеет только тривиальное решение $c_1 = c_2 = c_3 = 0$


***Вывод:***

Векторы $a_1 = (1, 8, -1)$, $a_2 = (-2, 3, 3)$, $a_3 = (4, -11, 9)$ линейно независимы

## Задание 2

При каких значениях параметра a система является линейно зависимой?  
Найдите решение без использования Python (вручную)

a1 = (1; 0; 2; -3),  
a2 = (2; 2; -4; a),  
a3 = (3; 1; 2; -5)

Для того чтобы система векторов $\mathbf{a}_1 = (1, 0, 2, -3)$, $\mathbf{a}_2 = (2, 2, -4, a)$, $\mathbf{a}_3 = (3, 1, 2, -5)$ была линейно зависимой, должно существовать нетривиальное решение уравнения:

$
c_1 \mathbf{a}_1 + c_2 \mathbf{a}_2 + c_3 \mathbf{a}_3 = \mathbf{0}
$

где $\mathbf{0}$ — нулевой вектор, и хотя бы один из коэффициентов $c_1, c_2, c_3$ отличен от нуля

Это эквивалентно тому, что матрица, составленная из этих векторов, должна иметь ранг, меньший числа векторов. 
Составим матрицу:

$
A = \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2 \\
-3 & a & -5
\end{bmatrix}
$

Чтобы найти условие линейной зависимости, рассмотрим определитель подматрицы из первых трех строк и столбцов:

$
\det \begin{bmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2
\end{bmatrix} = 1(2 \cdot 2 - 1 \cdot (-4)) - 2(0 \cdot 2 - 1 \cdot 2) + 3(0 \cdot (-4) - 2 \cdot 2)  $
$
= 1(4 + 4) - 2(0 - 2) + 3(0 - 4) = 8 + 4 - 12 = 0$


Поскольку определитель равен нулю, это значит, что векторы $\mathbf{a}_1, \mathbf{a}_2, \mathbf{a}_3$ уже линейно зависимы независимо от значения $a$

Следовательно, система линейно зависима для любого значения параметра $a$. Дополнительное выражение не требуется, так как уже три вектора линейно зависимы

## Задание 3
Выразите определитель матрицы. Найдите решение без использования Python (вручную)

\begin{vmatrix}
1 & -1 & a & -1 \\
-1 & -2 & b & -1 \\
-2 & 0 & c & 1 \\
0 & 1 & d & 0
\end{vmatrix}

Чтобы найти определитель этой матрицы $ \det(A) $, мы можем разложить его по элементам 4-ой строки
$
\det(A) = -0 \cdot \det(A_{41}) + 1 \cdot \det(A_{42}) - d \cdot \det(A_{43}) + 0 \cdot \det(A_{44})
$

где $A_{ij}$ — это дополнение элемента $a_{ij}$, то есть определитель матрицы, полученной из $A$ исключением $i$-й строки и $j$-го столбца.

Теперь вычислим каждый из этих определителей меньшего порядка $3 \times 3$:

$$ 

-0 \cdot 
\begin{vmatrix}
1 & a & -1\\
-2 & b & -1\\
0 & c & 1
\end{vmatrix}
+ 1 \cdot
\begin{vmatrix}
1 & a & -1\\
-1 & b & -1\\
-2 & c & 1
\end{vmatrix}
-d \cdot 
\begin{vmatrix}
1 & -1 & -1\\
-1 & -2 & -1\\
-2 & 0 & 1
\end{vmatrix}
+0 \cdot
\begin{vmatrix}
1 & -1 & a\\
-1 & -2 & b\\
-2 & 0 & c
\end{vmatrix} 
= 0 - b + 3a + 2c - d -1 + 0 = -b+3a+2c+d
$$

***Ответ:*** $-b+3a+2c+d$


## Задание 4
Найдите значения параметров a, b и c, при которых матрицы A и B являются обратными. Найдите решение без использования Python (вручную).

$A=\left(\begin{array}{ccc}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2 c
\end{array}\right), B=\left(\begin{array}{rrr}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{array}\right)$

Для того чтобы матрицы $ A $ и $ B $ были обратными, произведение $ AB $ и $ BA $ должно быть равно единичной матрице $ I $.

Вычислим произведение $ AB $:

$
AB = \begin{pmatrix}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2c
\end{pmatrix} 
\begin{pmatrix}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{pmatrix}
$

***Элементы первой строки:***

- Первый элемент: $ a \cdot (-4) + (-2) \cdot 0 + (-1) \cdot (-5) = -4a + 5 $
- Второй элемент: $ a \cdot 3 + (-2) \cdot 2 + (-1) \cdot (-1) =  3a - 3 $
- Третий элемент: $ a \cdot 5 + (-2) \cdot 3 + (-1) \cdot (-1) = 5a - 5 $

***Элементы второй строки:***

- Первый элемент: $ -15 \cdot (-4) + (b+20) \cdot 0 + 12 \cdot (-5) = 0 $
- Второй элемент: $ -15 \cdot 3 + (b+20) \cdot 2 + 12 \cdot (-1) = 2b - 17 $
- Третий элемент: $ -15 \cdot 5 + (b+20) \cdot 3 + 12 \cdot (-1) = 3b - 27 $

***Элементы третьей строки:***

- Первый элемент: $ 10 \cdot (-4) + (-19) \cdot 0 + 2c \cdot (-5) = -40 - 10c $
- Второй элемент: $ 10 \cdot 3 + (-19) \cdot 2 + 2c \cdot (-1) = -8 - 2c $
- Третий элемент: $ 10 \cdot 5 + (-19) \cdot 3 + 2c \cdot (-1) = -7 - 2c $

Для того чтобы $ AB = I $, матрица произведения должна быть равна единичной матрице:

$
\begin{pmatrix}
1 & 0 & 0 \\
0 & 1 & 0 \\
0 & 0 & 1
\end{pmatrix}
$

Теперь у нас есть следующие уравнения для нахождения $ a $, $ b $, $ c $:

1. $-4a + 5 = 1$
2. $3a - 3 = 0$
3. $5a - 5 = 0$
4. $2b - 17 = 1$
5. $3b - 27 = 0$
6. $-40 - 10c = 0$
7. $-8 - 2c = 0$
8. $-7 - 2c = 1$

Решим эти уравнения:

1. $3a - 3 = 0 \Rightarrow 3a = 3 \Rightarrow a = 1$

2. $2b - 17 = 1 \Rightarrow 2b = 18 \Rightarrow b = 9$

3. $-8 - 2c = 0 \Rightarrow -2c = 8 \Rightarrow c = -4$

То есть, значения параметров:

$
a = 1, \ b = 9, \ c = -4
$


## Задание 5
Найдите собственные векторы и собственные значения матрицы. Найдите решение без использования Python (вручную).

$\left(\begin{array}{rrr}4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4\end{array}\right)$

Для нахождения собственных значений и собственных векторов матрицы, нужно сначала найти собственные значения $\lambda$, решив характеристическое уравнение $\det(A - \lambda I) = 0$.

1. **Найдем характеристический многочлен:**

$
A - \lambda I = \begin{pmatrix} 4 - \lambda & 0 & -1 \\ 2 & 2 - \lambda & 1 \\ -1 & 0 & 4 - \lambda \end{pmatrix}
$

Найдем определитель:

$
\det(A - \lambda I) = \begin{vmatrix} 4 - \lambda & 0 & -1 \\ 2 & 2 - \lambda & 1 \\ -1 & 0 & 4 - \lambda \end{vmatrix} = - \lambda^3 +10\lambda^2-31\lambda+30 = 
-(\lambda-2)(\lambda^2-8\lambda+15)=-(\lambda-2)(\lambda-3)(\lambda-5)=0
$




Теперь найдем собственные векторы для данных собственных значений.

***$\lambda = 2$:***

Решаем $(A - 2I)\mathbf{v} = \mathbf{0}$:

$
A - 2I = \begin{pmatrix} 2 & 0 & -1 \\ 2 & 0 & 1 \\ -1 & 0 & 2 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
2x - z = 0 \\
2x + z = 0 \\
-x + 2z = 0
\end{cases}
$

Решение: $x = 0$, $z = 0$, $y$ свободный параметр, следовательно, $\mathbf{v}_1 = \begin{pmatrix} 0 \\ 1 \\ 0 \end{pmatrix}$.

***$\lambda = 3$:***

Решаем $(A - 3I)\mathbf{v} = \mathbf{0}$:

$
A - 3I = \begin{pmatrix} 1 & 0 & -1 \\ 2 & -1 & 1 \\ -1 & 0 & 1 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
x - z = 0 \\
2x - y + z = 0 \\
-x + z = 0
\end{cases}
$

Решение: $x = z$, $y = 3z$, следовательно, $\mathbf{v}_2 = \begin{pmatrix} 1 \\ 3 \\ 1 \end{pmatrix}$.

***$\lambda = 5$:***

Решаем $(A - 5I)\mathbf{v} = \mathbf{0}$:

$
A - 5I = \begin{pmatrix} -1 & 0 & -1 \\ 2 & -3 & 1 \\ -1 & 0 & -1 \end{pmatrix}
$

Система уравнений:

$
\begin{cases}
-x - z = 0 \\
2x - 3y + z = 0 \\
-x - z = 0
\end{cases}
$

Решение: $x = -z$, $y = -\frac{1}{3}z$, следовательно, $\mathbf{v}_3 = \begin{pmatrix} -1 \\ -\frac{1}{3} \\ 1 \end{pmatrix}$.

***Собственные значения и собственные векторы:***

1. $\lambda_1 = 2$, $\mathbf{v}_1 = \begin{pmatrix} 0 \\ 1 \\ 0 \end{pmatrix}$

2. $\lambda_2 = 3$, $\mathbf{v}_2 = \begin{pmatrix} 1 \\ 3 \\ 1 \end{pmatrix}$

3. $\lambda_3 = 5$, $\mathbf{v}_3 = \begin{pmatrix} -1 \\ -\frac{1}{3}\\ 1 \end{pmatrix}$

## Задание 6
Найдите сумму сингулярных чисел для матрицы:

$A=\left(\begin{array}{ccc}-2 & 0 & 1 \\ 0 & -2 & 1 \\ 1 & 1 & -1\end{array}\right)$

Найдите решение без использования Python (вручную).

***Найдём сумму сингулярных чисел для матрицы $A$***

Сингулярные числа матрицы $ A $ — это квадратные корни собственных значений матрицы $ A^T A $ (или $ AA^T $)

---

***Найдём $ A^T A $***
$
A^T A =
\begin{bmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{bmatrix}
\begin{bmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{bmatrix}
$

Первая строка:

$
(-2)(-2) + (0)(0) + (1)(1) = 5,
$

$
(-2)(0) + (0)(-2) + (1)(1) = 1,
$

$
(-2)(1) + (0)(1) + (1)(-1) = -3.
$

Вторая строка:

$
(0)(-2) + (-2)(0) + (1)(1) = 1,
$

$
(0)(0) + (-2)(-2) + (1)(1) = 5,
$

$
(0)(1) + (-2)(1) + (1)(-1) = -3.
$

Третья строка:

$
(1)(-2) + (1)(0) + (-1)(1) = -3,
$

$
(1)(0) + (1)(-2) + (-1)(1) =-3,
$

$
(1)(1) + (1)(1) + (-1)(-1) = 3.
$

$
A^T A =
\begin{bmatrix}
5 & 1 & -3 \\
1 & 5 & -3 \\
-3 & -3 & 3
\end{bmatrix}
$

---
***Найдём собственные значения $A^T A$***

Решаем характеристическое уравнение:

$
\det(A^T A - \lambda I) = 0.
$

$
A^T A - \lambda I =
\begin{bmatrix}
5 - \lambda & 1 & -3 \\
1 & 5 - \lambda & -3 \\
-3 & -3 & 3 - \lambda
\end{bmatrix} = -\lambda^3 + 13\lambda^2-36\lambda=-\lambda(\lambda^2-13\lambda+36) = -\lambda(\lambda-4)(\lambda-9)=0
$

$
\sigma_1 = \sqrt{0} = 0,
$
$
\sigma_2 = \sqrt{4} = 2,
$
$
\sigma_3 = \sqrt{9} = 3.
$

$
\sigma_1 + \sigma_2 + \sigma_3 = 0 + 2 + 3 = 5.
$

***Ответ:***
Сумма сингулярных чисел матрицы $A$ равна $5$.

# Часть 2: Задания, решаемые с помощью Python

Даны данные об аренде велосипедов.

Описание переменных:

- season: 1 - весна, 2 - лето, 3 - осень, 4 - зима
- yr: 0 - 2011, 1 - 2012
- mnth: месяц, от 1 до 12
- holiday: 0 - нет праздника, 1 - есть праздник
- weekday: день недели, от 0 до 6
- workingday: 0 - нерабочий день, 1 - рабочий день
- weathersit: оценка благоприятности погоды от 1 (чистый, ясный день) до 4 (ливень, туман)
- temp: температура в Цельсиях
- atemp: температура по ощущениям в Цельсиях
- hum: влажность
- windspeed(mph): скорость ветра в милях в час
- windspeed(ms): скорость ветра в метрах в секунду
- cnt: количество арендованных велосипедов (это целевой признак, его мы будем предсказывать)

## Задание 7
Загрузите данные. Изучите, есть ли в них пропущенные значения? Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [1]:
import pandas as pd

In [5]:
data = pd.read_csv(
    "data/asset_v1_SkillFactory+MIFIML_1sem+2024+type@asset+block@bikes_rent.xls"
)
data.head()

season  yr  mnth  holiday  weekday  workingday  weathersit       temp  \
0       1   0     1        0        6           0           2  14.110847   
1       1   0     1        0        0           0           2  14.902598   
2       1   0     1        0        1           1           1   8.050924   
3       1   0     1        0        2           1           1   8.200000   
4       1   0     1        0        3           1           1   9.305237   

      atemp      hum  windspeed(mph)  windspeed(ms)   cnt  
0  18.18125  80.5833       10.749882       4.805490   985  
1  17.68695  69.6087       16.652113       7.443949   801  
2   9.47025  43.7273       16.636703       7.437060  1349  
3  10.60610  59.0435       10.739832       4.800998  1562  
4  11.46350  43.6957       12.522300       5.597810  1600

In [6]:
data.isnull().sum()

season            0
yr                0
mnth              0
holiday           0
weekday           0
workingday        0
weathersit        0
temp              0
atemp             0
hum               0
windspeed(mph)    0
windspeed(ms)     0
cnt               0
dtype: int64

В данном датасете нет пропущенных значений по всем столбцам

## Задание 8
Исследуем данные на наличие линейной зависимости между столбцами. Найдите ранг матрицы данных. Сравните его с количеством столбцов, сделайте вывод. Если линейная зависимость есть - напишите, между какими столбцами и удалите один из них.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [10]:
import numpy as np

In [11]:
# рассчитаем корреляцию между числовыми столбцами
correlation_matrix = data.corr()
print('Корреляционная матрица: ')
correlation_matrix

Корреляционная матрица: 


season        yr      mnth   holiday   weekday  workingday  \
season          1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr             -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth            0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday        -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday        -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday      0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit      0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp            0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp           0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum             0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(mph) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
windspeed(ms)  -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt             0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

                weathersit      temp     atemp       hum  windspeed(mph)  \
season            0.019211  0.334315  0.342876  0.205445       -0.229046   
yr               -0.048727  0.047604  0.046106 -0.110651       -0.011817   
mnth              0.043528  0.220205  0.227459  0.222204       -0.207502   
holiday          -0.034627 -0.028556 -0.032507 -0.015937        0.006292   
weekday           0.031087 -0.000170 -0.007537 -0.052232        0.014282   
workingday        0.061200  0.052660  0.052182  0.024327       -0.018796   
weathersit        1.000000 -0.120602 -0.121583  0.591045        0.039511   
temp             -0.120602  1.000000  0.991702  0.126963       -0.157944   
atemp            -0.121583  0.991702  1.000000  0.139988       -0.183643   
hum               0.591045  0.126963  0.139988  1.000000       -0.248489   
windspeed(mph)    0.039511 -0.157944 -0.183643 -0.248489        1.000000   
windspeed(ms)     0.039511 -0.157944 -0.183643 -0.248489        1.000000   
cnt              -0.297391  0.627494  0.631066 -0.100659       -0.234545   

                windspeed(ms)       cnt  
season              -0.229046  0.406100  
yr                  -0.011817  0.566710  
mnth                -0.207502  0.279977  
holiday              0.006292 -0.068348  
weekday              0.014282  0.067443  
workingday          -0.018796  0.061156  
weathersit           0.039511 -0.297391  
temp                -0.157944  0.627494  
atemp               -0.183643  0.631066  
hum                 -0.248489 -0.100659  
windspeed(mph)       1.000000 -0.234545  
windspeed(ms)        1.000000 -0.234545  
cnt                 -0.234545  1.000000

In [12]:
# найдем ранг матрицы
matrix_rank = np.linalg.matrix_rank(data.select_dtypes(include=[np.number]))
print("Ранг матрицы:", matrix_rank)


Ранг матрицы: 12


In [13]:
# cравним ранг с количеством столбцов
num_columns = data.select_dtypes(include=[np.number]).shape[1]
print("Количество столбцов:", num_columns)

Количество столбцов: 13


In [20]:
# вывод по линейной зависимости
if matrix_rank < num_columns:

    print("Есть линейная зависимость между некоторыми столбцами")

    # Находим пары столбцов с высокой корреляцией (выше 0.8)
    high_correlation = correlation_matrix[
        (correlation_matrix > 0.8) & (correlation_matrix < 1.0)
    ]
    print("\nСтолбцы с высокой корреляцией:")
    display(high_correlation)

else:
    print("\nНет линейной зависимости между столбцами")

Есть линейная зависимость между некоторыми столбцами

Столбцы с высокой корреляцией:


season  yr     mnth  holiday  weekday  workingday  \
season              NaN NaN  0.83144      NaN      NaN         NaN   
yr                  NaN NaN      NaN      NaN      NaN         NaN   
mnth            0.83144 NaN      NaN      NaN      NaN         NaN   
holiday             NaN NaN      NaN      NaN      NaN         NaN   
weekday             NaN NaN      NaN      NaN      NaN         NaN   
workingday          NaN NaN      NaN      NaN      NaN         NaN   
weathersit          NaN NaN      NaN      NaN      NaN         NaN   
temp                NaN NaN      NaN      NaN      NaN         NaN   
atemp               NaN NaN      NaN      NaN      NaN         NaN   
hum                 NaN NaN      NaN      NaN      NaN         NaN   
windspeed(mph)      NaN NaN      NaN      NaN      NaN         NaN   
windspeed(ms)       NaN NaN      NaN      NaN      NaN         NaN   
cnt                 NaN NaN      NaN      NaN      NaN         NaN   

                weathersit      temp     atemp  hum  windspeed(mph)  \
season                 NaN       NaN       NaN  NaN             NaN   
yr                     NaN       NaN       NaN  NaN             NaN   
mnth                   NaN       NaN       NaN  NaN             NaN   
holiday                NaN       NaN       NaN  NaN             NaN   
weekday                NaN       NaN       NaN  NaN             NaN   
workingday             NaN       NaN       NaN  NaN             NaN   
weathersit             NaN       NaN       NaN  NaN             NaN   
temp                   NaN       NaN  0.991702  NaN             NaN   
atemp                  NaN  0.991702       NaN  NaN             NaN   
hum                    NaN       NaN       NaN  NaN             NaN   
windspeed(mph)         NaN       NaN       NaN  NaN             NaN   
windspeed(ms)          NaN       NaN       NaN  NaN             1.0   
cnt                    NaN       NaN       NaN  NaN             NaN   

                windspeed(ms)  cnt  
season                    NaN  NaN  
yr                        NaN  NaN  
mnth                      NaN  NaN  
holiday                   NaN  NaN  
weekday                   NaN  NaN  
workingday                NaN  NaN  
weathersit                NaN  NaN  
temp                      NaN  NaN  
atemp                     NaN  NaN  
hum                       NaN  NaN  
windspeed(mph)            1.0  NaN  
windspeed(ms)             NaN  NaN  
cnt                       NaN  NaN

На основе полученной корреляционной матрицы видно, что значения корреляции между некоторыми столбцами очень высоки

- temp и atemp: корреляция 0.991702 — зависимость между температурой и температурой по ощущениям.
- windspeed(mph) и windspeed(ms): корреляция 1.0 — полная зависимость, так как скорость ветра измеряется в двух единицах. 

Удалим столбец windspeed(mph)

In [21]:
del data["windspeed(mph)"]

## Задание 9
Выведите корреляционную матрицу оставшихся переменных.

Какая переменная сильнее всего коррелирует с целевой?
Есть ли мультиколлинеарность? Между какими переменными? Если есть корреляция выше 0.9 - оставьте только 1 признак. Объясните выбор
Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [29]:
data.corr()

season        yr      mnth   holiday   weekday  workingday  \
season         1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr            -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth           0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday       -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday       -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday     0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit     0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp           0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp          0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum            0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(ms) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt            0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

               weathersit      temp     atemp       hum  windspeed(ms)  \
season           0.019211  0.334315  0.342876  0.205445      -0.229046   
yr              -0.048727  0.047604  0.046106 -0.110651      -0.011817   
mnth             0.043528  0.220205  0.227459  0.222204      -0.207502   
holiday         -0.034627 -0.028556 -0.032507 -0.015937       0.006292   
weekday          0.031087 -0.000170 -0.007537 -0.052232       0.014282   
workingday       0.061200  0.052660  0.052182  0.024327      -0.018796   
weathersit       1.000000 -0.120602 -0.121583  0.591045       0.039511   
temp            -0.120602  1.000000  0.991702  0.126963      -0.157944   
atemp           -0.121583  0.991702  1.000000  0.139988      -0.183643   
hum              0.591045  0.126963  0.139988  1.000000      -0.248489   
windspeed(ms)    0.039511 -0.157944 -0.183643 -0.248489       1.000000   
cnt             -0.297391  0.627494  0.631066 -0.100659      -0.234545   

                    cnt  
season         0.406100  
yr             0.566710  
mnth           0.279977  
holiday       -0.068348  
weekday        0.067443  
workingday     0.061156  
weathersit    -0.297391  
temp           0.627494  
atemp          0.631066  
hum           -0.100659  
windspeed(ms) -0.234545  
cnt            1.000000

In [30]:
data.corr()["cnt"].to_frame().sort_values(by="cnt")

cnt
weathersit    -0.297391
windspeed(ms) -0.234545
hum           -0.100659
holiday       -0.068348
workingday     0.061156
weekday        0.067443
mnth           0.279977
season         0.406100
yr             0.566710
temp           0.627494
atemp          0.631066
cnt            1.000000

- Самая сильная корреляция с целевой переменной cnt: это столбцы atemp (0.631066) и temp (0.627494). Обе эти переменные связаны с температурой, причем atemp (температура по ощущениям) более сильно коррелирует с количеством арендованных велосипедов

- Также между переменными temp и atemp высокая корреляция (0.991702), что означает, что эти два признака сильно линейно зависимы. Такая ситуация является проявлением мультиколлинеарности, и чтобы ее устранить, оставим только один из этих признаков

Удалим температуру по ощущениям

In [31]:
del data["atemp"]

## Задание 10
Разделите данные на предикторы и целевую переменную. Обучите модель с помощью классической линейной регрессии. Линейную регрессию необходимо реализовать без готовых библиотек.

Оцените качество модели с помощью хотя бы двух метрик качества.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [32]:
X = data.drop(columns=["cnt"])  # все переменные, кроме целевой переменной
y = data["cnt"]  # целевая переменная

In [33]:
# столбец единичных значений для смещения (bias)
X = np.c_[np.ones(X.shape[0]), X]

X = np.array(X)
y = np.array(y)

# находим коэффициенты линейной регрессии
beta = np.linalg.inv(X.T @ X) @ X.T @ y

In [34]:
# получаем прогнозы
y_pred = X @ beta

In [35]:
# Среднеквадратичная ошибка
mse = np.mean((y - y_pred) ** 2)

# Коэффициент детерминации R^2
ss_total = np.sum((y - np.mean(y)) ** 2)
ss_residual = np.sum((y - y_pred) ** 2)
r2 = 1 - (ss_residual / ss_total)

print(f"MSE: {mse}")
print(f"R^2: {r2}")

MSE: 753878.6752929664
R^2: 0.7988398641466421
